Design/Algorithm_ease_of_use.md

```diff
@@ -0,0 +1,100 @@
+if ((m_outputEW) && (tolerance > 0.)) {
+  g_log.information() << "running CompressEvents(Tolerance=" << tolerance
+                      << ")\n";
+  API::IAlgorithm_sptr compressAlg = createChildAlgorithm("CompressEvents");
+  compressAlg->setProperty("InputWorkspace", m_outputEW);
+  compressAlg->setProperty("OutputWorkspace", m_outputEW);
+  compressAlg->setProperty("Tolerance", tolerance);
+  compressAlg->executeAsChildAlg();
+  m_outputEW = compressAlg->getProperty("OutputWorkspace");
+  m_outputW = boost::dynamic_pointer_cast<MatrixWorkspace>(m_outputEW);
+}
+m_progress->report();
+```
+
+This boiler-plate code exists (sometimes with copy/paste errors) in
+many algorithms.
+
+Alternative 1
+=============
+
+Add specific code into algorithms that should be simpler to call. This
+would be strongly typed methods on the classes themselves. The code in
+the header would be:
+
+```C
+static
+EventWorkspace_sptr CompressEvents::executeAsChildAlg(EventWorkspace_sptr input,
+                                                      EventWorkspace_sptr output,
+                                                      double tolerance=1.e-5);
+```
+
+The return is the pointer to the resultant output workspace. This is
+required because we use copy-on-write (cow) pointers. In the example
+above its usage would be
+
+```C
+double tolerance = getProperty("CompressTolerance");
+m_outputEW = boost::dynamic_pointer_cast<EventWorkspace>(m_outputW);
+if ((m_outputEW) && (tolerance > 0.)) {
+  g_log.information() << "running CompressEvents(Tolerance=" << tolerance
+                      << ")\n";
+  m_outputEW = CompressEvents::executeAsChildAlg(m_outputEW, m_outputEW, tolerance);
+  m_outputW = boost::dynamic_pointer_cast<MatrixWorkspace>(m_outputEW);
+}
+m_progress->report();
+```
+
+While this is much cleaner in client code, every algorithm would need
+to modified to add this functionality.
+
+Alternative 2
+=============
+
+Initializer lists
+([cplusplus](http://www.cplusplus.com/reference/initializer_list/initializer_list/))
+or variadic templates
+([cplusplus](http://www.cplusplus.com/articles/EhvU7k9E/) ,
+[msdn](https://msdn.microsoft.com/en-us/library/dn439779.aspx)), might
+do the job well. What this would do internally is call `setProperty(string &, Type &)` for each pair in the list then call `executateAsChild()`. The details of how this would exactly work are a bit fuzzy, but the usage would be:
+
+```C
+double tolerance = getProperty("CompressTolerance");
+m_outputEW = boost::dynamic_pointer_cast<EventWorkspace>(m_outputW);
+if ((m_outputEW) && (tolerance > 0.)) {
+  g_log.information() << "running CompressEvents(Tolerance=" << tolerance
+                      << ")\n";
+  API::IAlgorithm_sptr compressAlg = createChildAlgorithm("CompressEvents");
+  compressAlg->executeAsChildAlg({"InputWorkspace", m_outputEW},
+                           {"OutputWorkspace", m_outputEW},
+                           {"Tolerance", tolerance});
```

Comment:
I like the direction in which Alternative 2/3? is going.
Now what will code complexity tools say about this? They will disklike the type of calls that we will have here, with many properties passed in a single method call.
Maybe algorithms should have an input dictionary and an output dictionary of properties?